File: android/app/src/main/java/net/quakemesh/android/transport/LanBeacon.kt
// Copyright (c) 2026 John Dovey <dovey.john@gmail.com>
//
// Changelog:
//   0.0.18 - LAN multicast presence beacons (matches core/lanbeacon).
//   0.0.19 - optional lan_context on node beacons.

package net.quakemesh.android.transport

import org.json.JSONObject

/** Wire format shared with QuakeMeshHub [core/lanbeacon]. */
object LanBeacon {
    const val KIND_HUB = "hub"
    const val KIND_NODE = "node"

    private val PREFIX = byteArrayOf('Q'.code.toByte(), 'M'.code.toByte(), 'L'.code.toByte(), 'B'.code.toByte(), 1)

    data class HubBeacon(
        val nodeId: String,
        val heartbeatPort: Int,
        val ogmPort: Int,
    )

    fun isBeacon(payload: ByteArray): Boolean {
        if (payload.size < PREFIX.size) return false
        return PREFIX.indices.all { payload[it] == PREFIX[it] }
    }

    data class LanContext(
        val gatewayIp: String,
        val localIp: String? = null,
        val ssid: String? = null,
        val bssid: String? = null,
    )

    fun encodeNode(
        nodeId: String,
        handle: String? = null,
        homeLat: Double? = null,
        homeLon: Double? = null,
        lat: Double?,
        lon: Double?,
        accuracyM: Double?,
        lan: LanContext? = null,
    ): ByteArray {
        val json = JSONObject()
            .put("v", 1)
            .put("kind", KIND_NODE)
            .put("node_id", nodeId)
        if (!handle.isNullOrBlank()) {
            json.put("handle", handle.trim())
        }
        if (homeLat != null && homeLon != null) {
            json.put("home_lat", homeLat).put("home_lon", homeLon)
        }
        if (lat != null && lon != null) {
            json.put("lat", lat).put("lon", lon)
            if (accuracyM != null) {
                json.put("accuracy_m", accuracyM)
            }
        }
        lan?.let { putLanContext(json, it) }
        return PREFIX + json.toString().toByteArray(Charsets.UTF_8)
    }

    private fun putLanContext(json: JSONObject, lan: LanContext) {
        val ctx = JSONObject().put("gateway_ip", lan.gatewayIp)
        lan.localIp?.let { ctx.put("local_ip", it) }
        lan.ssid?.let { ctx.put("ssid", it) }
        lan.bssid?.let { ctx.put("bssid", it) }
        json.put("lan_context", ctx)
    }

    fun lanContextFromJson(obj: JSONObject?): LanContext? {
        if (obj == null) return null
        val gateway = obj.optString("gateway_ip")
        if (gateway.isBlank()) return null
        return LanContext(
            gatewayIp = gateway,
            localIp = obj.optString("local_ip").takeIf { it.isNotBlank() },
            ssid = obj.optString("ssid").takeIf { it.isNotBlank() },
            bssid = obj.optString("bssid").takeIf { it.isNotBlank() },
        )
    }

    fun decodeHub(payload: ByteArray): HubBeacon? {
        if (!isBeacon(payload)) return null
        val json = JSONObject(String(payload, PREFIX.size, payload.size - PREFIX.size, Charsets.UTF_8))
        if (json.optString("kind") != KIND_HUB) return null
        val nodeId = json.optString("node_id")
        if (nodeId.isBlank()) return null
        val heartbeatPort = json.optInt("heartbeat_port", 0)
        val ogmPort = json.optInt("ogm_port", 0)
        if (heartbeatPort <= 0 || ogmPort <= 0) return null
        return HubBeacon(nodeId, heartbeatPort, ogmPort)
    }

    data class NodeBeacon(
        val nodeId: String,
        val handle: String?,
        val homeLat: Double?,
        val homeLon: Double?,
        val lat: Double?,
        val lon: Double?,
    )

    fun decodeNode(payload: ByteArray): NodeBeacon? {
        if (!isBeacon(payload)) return null
        val json = JSONObject(String(payload, PREFIX.size, payload.size - PREFIX.size, Charsets.UTF_8))
        if (json.optString("kind") != KIND_NODE) return null
        val nodeId = json.optString("node_id")
        if (nodeId.isBlank()) return null
        val handle = json.optString("handle").takeIf { it.isNotBlank() }
        val homeLat = if (json.has("home_lat")) json.optDouble("home_lat") else null
        val homeLon = if (json.has("home_lon")) json.optDouble("home_lon") else null
        val lat = if (json.has("lat")) json.optDouble("lat") else null
        val lon = if (json.has("lon")) json.optDouble("lon") else null
        return NodeBeacon(nodeId, handle, homeLat, homeLon, lat, lon)
    }
}

File: core/lanbeacon/lanbeacon.go
// Copyright (c) 2026 John Dovey <dovey.john@gmail.com>
//
// Changelog:
//   0.0.18 - LAN multicast presence beacons for hub/node discovery.
//   0.0.19 - optional lan_context on hub/node beacons.

// Package lanbeacon defines the JSON UDP multicast format used on the
// connected Wi-Fi LAN (see "LAN discovery" in /plan.md).
package lanbeacon

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"

	"github.com/JohnDovey/QuakeMesh/core/lancontext"
)

const (
	// MulticastGroup is the LAN discovery multicast address.
	MulticastGroup = "239.255.42.99"
	// MulticastPort is shared by hub and node presence beacons.
	MulticastPort = 47223
)

const (
	magic0 = 'Q'
	magic1 = 'M'
	magic2 = 'L'
	magic3 = 'B'
	version = 1
)

const (
	KindHub  = "hub"
	KindNode = "node"
)

// Message is a presence beacon payload.
type Message struct {
	V             int                 `json:"v"`
	Kind          string              `json:"kind"`
	NodeID        string              `json:"node_id"`
	Handle        string              `json:"handle,omitempty"`
	HomeLat       *float64            `json:"home_lat,omitempty"`
	HomeLon       *float64            `json:"home_lon,omitempty"`
	HeartbeatPort int                 `json:"heartbeat_port,omitempty"`
	OGMPort       int                 `json:"ogm_port,omitempty"`
	Lat           *float64            `json:"lat,omitempty"`
	Lon           *float64            `json:"lon,omitempty"`
	AccuracyM     *float64            `json:"accuracy_m,omitempty"`
	LanContext    *lancontext.Context `json:"lan_context,omitempty"`
}

// Encode marshals a beacon with the QuakeMesh LAN prefix.
func Encode(msg Message) ([]byte, error) {
	if msg.V == 0 {
		msg.V = 1
	}
	if msg.Kind == "" {
		return nil, errors.New("lanbeacon: kind required")
	}
	if msg.NodeID == "" {
		return nil, errors.New("lanbeacon: node_id required")
	}
	body, err := json.Marshal(msg)
	if err != nil {
		return nil, err
	}
	out := make([]byte, 0, 5+len(body))
	out = append(out, magic0, magic1, magic2, magic3, version)
	out = append(out, body...)
	return out, nil
}

// Decode parses a beacon datagram. Returns false when the payload is not a
// QuakeMesh LAN beacon (e.g. a mesh frame on the same port).
func Decode(payload []byte) (Message, bool, error) {
	if len(payload) < 6 || payload[0] != magic0 || payload[1] != magic1 ||
		payload[2] != magic2 || payload[3] != magic3 || payload[4] != version {
		return Message{}, false, nil
	}
	var msg Message
	if err := json.Unmarshal(payload[5:], &msg); err != nil {
		return Message{}, true, err
	}
	if msg.Kind == "" || msg.NodeID == "" {
		return Message{}, true, errors.New("lanbeacon: invalid message")
	}
	return msg, true, nil
}

// IsBeacon reports whether payload starts with the QuakeMesh LAN prefix.
func IsBeacon(payload []byte) bool {
	_, ok, _ := Decode(payload)
	return ok
}

// HubBeacon builds a hub announcement.
func HubBeacon(nodeID string, heartbeatPort, ogmPort int, lan *lancontext.Context) ([]byte, error) {
	if heartbeatPort <= 0 || ogmPort <= 0 {
		return nil, fmt.Errorf("lanbeacon: invalid hub ports heartbeat=%d ogm=%d", heartbeatPort, ogmPort)
	}
	return Encode(Message{
		V:             1,
		Kind:          KindHub,
		NodeID:        nodeID,
		HeartbeatPort: heartbeatPort,
		OGMPort:       ogmPort,
		LanContext:    lan,
	})
}

// NodeBeacon builds a node announcement.
func NodeBeacon(nodeID string, handle string, homeLat, homeLon, lat, lon, accuracyM *float64, lan *lancontext.Context) ([]byte, error) {
	return Encode(Message{
		V:          1,
		Kind:       KindNode,
		NodeID:     nodeID,
		Handle:     handle,
		HomeLat:    homeLat,
		HomeLon:    homeLon,
		Lat:        lat,
		Lon:        lon,
		AccuracyM:  accuracyM,
		LanContext: lan,
	})
}

// Prefix returns the wire prefix bytes (for tests).
func Prefix() []byte {
	return bytes.Clone([]byte{magic0, magic1, magic2, magic3, version})
}
